addons/project/models/project_task.py

```diff
@@ -1273,40 +1216,49 @@ def write(self, vals):
 
         # sends an email to the 'Task Creation' subtype subscribers
         # When project_id is changed
+        project_link_per_task_id = {}
         if vals.get('project_id'):
             project = self.env['project.project'].browse(vals.get('project_id'))
             notification_subtype_id = self.env['ir.model.data']._xmlid_to_res_id('project.mt_project_task_new')
             partner_ids = project.message_follower_ids.filtered(lambda follower: notification_subtype_id in follower.subtype_ids.ids).partner_id.ids
             if partner_ids:
-                link_per_project_id = {}
-                for task in self:
-                    if task.project_id:
-                        project_link = link_per_project_id.get(task.project_id.id)
-                        if not project_link:
-                            project_link = link_per_project_id[task.project_id.id] = task.project_id._get_html_link(title=task.project_id.display_name)
-                        project_link_per_task_id[task.id] = project_link
+                link_per_project_id = {
+                    project.id: project._get_html_link(title=project.display_name)
+                    for project in self.project_id
+                }
+                project_link_per_task_id = {
+                    task.id: link_per_project_id[task.project_id.id]
+                    for task in self
+                    if task.project_id
+                }
```

Comment:
It does not seem related to the purpose of the commit. 🤔 

I am a bit curious, why the previous code was not good? Does it not better to do one for loop when we can? 🤔 

TBH, I am against to change that code if you think it is better to use multiple list comprehensions instead of doing a basic for loop. 🤷🏼‍♂️

Reply:
I will rollback this change.

It is just a matter of functional vs imperative style.
I had an error at some point and needed to investigate where this comes from and found that harder to read than "gather all project links and then assign each task id to a project".